File: data/derived_cot/rq3_prompting/results/translation/qwen/py_to_cpp/code_output/BoyerMooreSearch.cpp
#include <vector>
#include <string>

class BoyerMooreSearch {
private:
    std::string text;
    std::string pattern;
    int textLen;
    int patLen;

    int match_in_pattern(char c) {
        for (int i = patLen - 1; i >= 0; i--) {
            if (c == pattern[i]) {
                return i;
            }
        }
        return -1;
    }

    int mismatch_in_text(int currentPos) {
        for (int i = patLen - 1; i >= 0; i--) {
            if (pattern[i] != text[currentPos + i]) {
                return currentPos + i;
            }
        }
        return -1;
    }

public:
    BoyerMooreSearch(const std::string& text, const std::string& pattern)
        : text(text), pattern(pattern), textLen(text.length()), patLen(pattern.length()) {}

    std::vector<int> bad_character_heuristic() {
        std::vector<int> positions;
        for (int i = 0; i <= textLen - patLen; i++) {
            int mismatch_index = mismatch_in_text(i);
            if (mismatch_index == -1) {
                positions.push_back(i);
            } else {
                int match_index = match_in_pattern(text[mismatch_index]);
                i = mismatch_index - match_index;
            }
        }
        return positions;
    }
};

File: data/derived_cot/rq3_prompting/results/translation/glm-5.1/java_to_py/code_output/BoyerMooreSearch.py
class BoyerMooreSearch:
    def __init__(self, text, pattern):
        self.text = text
        self.pattern = pattern
        self.textLen = len(text)
        self.patLen = len(pattern)

    def matchInPattern(self, ch):
        return self.pattern.rfind(ch)

    def mismatchInText(self, currentPos):
        for i in range(self.patLen - 1, -1, -1):
            if self.pattern[i] != self.text[currentPos + i]:
                return currentPos + i
        return -1

    def badCharacterHeuristic(self):
        positions = []
        i = 0

        badCharHeuristic = {}
        for j in range(self.patLen):
            badCharHeuristic[self.pattern[j]] = j

        if self.patLen == 0:
            return list(range(self.textLen + 1))

        while i <= self.textLen - self.patLen:
            mismatchIndex = self.mismatchInText(i)
            if mismatchIndex == -1:
                positions.append(i)
                i += self.patLen
            else:
                mismatchChar = self.text[mismatchIndex]
                matchIndex = badCharHeuristic.get(mismatchChar, -1)
                if matchIndex >= 0:
                    i += max(1, mismatchIndex - i - matchIndex)
                else:
                    i += mismatchIndex - i + 1
        return positions